### PR TITLE
PMM-7412 implement api for getting logs

### PR DESCRIPTION
**Pull request by jprukner**
https://jira.percona.com/browse/PMM-7412

Implement api for getting logs - latest 1000 lines of logs, it consists of cluster's containers logs and events from its pods.
I included the changes to github actions to make tests run - they won't run with latest minikube.

### CODE REVIEW

File: service/k8sclient/k8sclient.go [hunk 1]
```diff
@@ -746,7 +747,7 @@ func (c *K8sClient) CreatePSMDBCluster(ctx context.Context, params *PSMDBParams)
 		return errors.Wrap(err, "cannot create secret for PXC")
 	}
 
-	affinity := &psmdb.PodAffinity{}
+	affinity := new(psmdb.PodAffinity)
```
Comment (palash25) [inline service/k8sclient/k8sclient.go:750]: what's the reason for this change? we usually don't use `new` in Go unless we really need to. So if there is no special need for this, I would suggest not using new so that we can keep declarations consistent throughout our codebases

Comment (jprukner) [reply service/k8sclient/k8sclient.go:750]: `bin/go-consistent -pedantic -exclude "tests" ./...` kept failing on this. I was thinking: What's the point of having tests that fail and we ignore them failing. So I changed it.

Comment (jprukner) [reply service/k8sclient/k8sclient.go:750]: `zero value ptr alloc: use new(T) for *T allocation` is an error one gets.

Comment (AlekSi) [reply service/k8sclient/k8sclient.go:750]: https://github.com/percona-platform/dbaas-controller/pull/78

If it works now, I think we should remove `-exclude "tests"` and then update _tests_ (despite go-consistent complaining about non-testing code – it picks the most common / consistent variant, but we want consistency avoid `new`). But let's do that in a separate PR

Comment (jprukner) [reply service/k8sclient/k8sclient.go:750]: I will remove this usage of `new` then.

File: utils/servers/grpc.go [hunk 2]
```diff
@@ -145,3 +158,35 @@ func (s *grpcServer) Run(ctx context.Context) {
 
 	<-stopped
 }
+
+// injectK8sClient tries to store k8sclient into context. It inspects request's
+// field KubeAuth. If it contains filed Kubeconfig, k8sclient is injected.
+func injectK8sClient(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
```
Comment (BupycHuk) [inline utils/servers/grpc.go:164]: looks like complication of simple code, isn't it?

File: cmd/dbaas-controller/main.go [hunk 3]
```diff
@@ -26,8 +26,10 @@ import (
 	"google.golang.org/grpc/grpclog"
 	"gopkg.in/alecthomas/kingpin.v2"
 
-	_ "github.com/percona-platform/dbaas-controller/catalog" // load messages.
+	// load messages.
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply cmd/dbaas-controller/main.go:29]: I know that it was done because of golint, but he is wrong

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: utils/servers/grpc.go (hunk 2 above)
Comment (jprukner) [reply utils/servers/grpc.go:164]: What would you recommend? I can remove this and create the client the same way we do it in other places.
I thought we could use this for every call in the future. So we don't have to call `defer client.Cleanup()` in every API call.

File: cmd/dbaas-controller/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jprukner) [reply cmd/dbaas-controller/main.go]: Done

File: service/logs/logs.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// Thanks to source interface we can get logs from different sources.
+type source interface {
+	getLogs(ctx context.Context, client *k8sclient.K8sClient, clusterName string) ([]*controllerv1beta1.Logs, error)
+}
+
+// allLogsSource implements source interface, it gets all logs from all
+// cluster's containers. It also gets events out of all cluster's pods.
+type allLogsSource struct{}
```
[the remark this answers is not in the document]
Comment (jprukner) [reply service/logs/logs.go]: Done

File: service/logs/logs.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+// getLogs gets all logs from all cluster's containers and events from all pods.
+func (a allLogsSource) getLogs(ctx context.Context, client *k8sclient.K8sClient, clusterName string) ([]*controllerv1beta1.Logs, error) {
```
[the remark this answers is not in the document]
Comment (jprukner) [reply service/logs/logs.go]: Done

File: utils/servers/grpc.go (hunk 2 above)
Comment (jprukner) [reply utils/servers/grpc.go]: Done

**Comment by jprukner** [comment]
Rebased, conflicts solved.

**Review by askomorokhov — APPROVED** [review]
Overall looks good to me. 👍 

It will be nice to have API test for this as well https://github.com/percona-platform/dbaas-controller/tree/main/tests/kubernetes.

Have you tested it under the big load? May this functionality introduce any performance issues?

**Review by percona-csalguero — APPROVED** [review]

**Comment by jprukner** [comment]
> Overall looks good to me. +1
> 
> It will be nice to have API test for this as well https://github.com/percona-platform/dbaas-controller/tree/main/tests/kubernetes.
> 
:+1: 
> Have you tested it under the big load? May this functionality introduce any performance issues?
>
I haven't tested it under high load. But will try it.

File: service/logs/all_logs_source.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+	controllerv1beta1 "github.com/percona-platform/dbaas-api/gen/controller"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/percona-platform/dbaas-controller/service/k8sclient"
+)
+
+// overallLinesLimit defines how many lines of logs we should return upon
```
Comment (BupycHuk) [inline service/logs/all_logs_source.go:32]: Do we return last 1000 lines or first 1000 lines?

Comment (jprukner) [reply service/logs/all_logs_source.go:32]: last 1000 lines

Comment (jprukner) [reply service/logs/all_logs_source.go:32]: I will modify the comment so it reflects the fact,

File: service/logs/all_logs_source.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	sum := 0
+	for _, count := range counts {
+		sum += count
+	}
+	return sum
+}
+
+// limitLines limits each entry's logs lines count in the way the overall sum of
+// all log lines is equal to given limit.
+func limitLines(logs []*controllerv1beta1.Logs, limit int) {
```
Comment (BupycHuk) [inline service/logs/all_logs_source.go:88]: I'm pretty sure there should be easier way to calculate the value.
Which won't be O(n2).
What about this logic?
```
counts := make(map[int]struct{})
fillCounts()
for lastSum < limit && len(counts) > 0 {
    a := (limit-lastSum)/len(counts)
    idx, logLines := findItemWithMinimumLog(logs)
    if logLines < a {
        delete(counts[idx])
        lastSum  += logLines
    } else {
        for i := range counts {
            logs[i].Logs =  logs[i].Logs[len(logs[i].Logs)-(lastSum + a)]
        }
        break
    }
}
```

Comment (BupycHuk) [inline service/logs/all_logs_source.go:88]: let's write some tests for `limitLines`.

Comment (jprukner) [reply service/logs/all_logs_source.go:88]: I get the theme of your logic. Let me try to modify it.

Comment (BupycHuk) [reply service/logs/all_logs_source.go:88]: let's write test first and then modify :)

Comment (jprukner) [reply service/logs/all_logs_source.go:80]: Tests were written.

Comment (jprukner) [reply service/logs/all_logs_source.go:80]: Tests were written.

**Comment by jprukner** [comment]
@BupycHuk I am sorry for force pushing, I accidentally pushed commits to this branch and had to reset it --hard to previous state.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: service/k8sclient/k8sclient.go [hunk 8]
```diff
@@ -1195,3 +1196,47 @@ func (c *K8sClient) checkOperatorStatus(installedVersions []string, expectedAPIV
 	}
 	return OperatorStatusNotInstalled
 }
+
+// GetClusterPods returns list of cluster's pods. It finds them by given cluster
+// name.
+func (c *K8sClient) GetClusterPods(ctx context.Context, clusterName string) (*common.PodList, error) {
+	list := &common.PodList{}
```
Comment (BupycHuk) [inline service/k8sclient/k8sclient.go:1203]: looks like it shouldn't be pointer here or in line 1208 we shouldn't use &

File: service/k8sclient/k8sclient_test.go [hunk 9]
```diff
@@ -94,6 +102,55 @@ func TestK8sClient(t *testing.T) {
...
+				},
+				{
+					name:       name + "-pxc-0",
+					containers: []string{"pxc", "pmm-client"},
+				},
+			}
+			for _, ppod := range pods.Items {
+				var foundPod pod
+				var found bool
+				for _, expectedPod := range expectedPods {
```
[the remark this answers is not in the document]
Comment (jprukner) [reply service/k8sclient/k8sclient_test.go:122]: Not sure where you expect me to use `assert.Condition`. I need the if because I am searching for the pod name in list of expected pod names.

File: service/logs/all_logs_source_test.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+			},
+		},
+	}
+	for _, tc := range testCases {
+		limitLines(tc.input, tc.limit)
+		assert.Truef(t, equal(tc.input, tc.expected), "expected %v\ngot %v", tc.expected, tc.input)
+	}
+}
+
+func equal(a, b []*controllerv1beta1.Logs) bool {
```
[the remark this answers is not in the document]
Comment (jprukner) [reply service/logs/all_logs_source_test.go]: I removed the func equal and used assert.Equal instead.

File: service/k8sclient/k8sclient.go (hunk 8 above)
Comment (jprukner) [reply service/k8sclient/k8sclient.go:1203]: Fixed

File: service/k8sclient/k8sclient_test.go [hunk 11]
```diff
@@ -118,35 +118,32 @@ func TestK8sClient(t *testing.T) {
...
+						return false
+					}, "container name '%s' was not expected", container.Name)
+
 					logs, err := client.GetLogs(ctx, ppod.Name, container.Name)
 					require.NoError(t, err, "failed to get logs")
 					assert.Greater(t, len(logs), 0)
 					for _, l := range logs {
 						assert.False(t, strings.Contains(l, "\n"), "new lines should have been removed")
 					}
-					if container.Name == "" {
```
Comment (jprukner) [inline service/k8sclient/k8sclient_test.go]: Right now when limit of lines is set to 1000 lines, this check should pass every time. I removed it to make it not fail if we decrease the limit substantially in the future.

File: service/k8sclient/k8sclient_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jprukner) [reply service/k8sclient/k8sclient_test.go]: I use `assert.Condition` now.

File: service/k8sclient/k8sclient.go [hunk 12]
```diff
@@ -1214,3 +1215,47 @@ func (c *K8sClient) checkOperatorStatus(installedVersions []string, expectedAPIV
...
```
Comment (BupycHuk) [inline service/k8sclient/k8sclient.go:1222]: let's keep using `new`

File: service/k8sclient/k8sclient_test.go [hunk 13]
```diff
@@ -110,6 +118,52 @@ func TestK8sClient(t *testing.T) {
...
+					containers: []string{"pmm-client", "proxysql", "pxc-monit", "proxysql-monit"},
+				},
+				{
+					name:       name + "-pxc-0",
+					containers: []string{"pxc", "pmm-client", "pxc-init"},
+				},
+			}
+			for _, ppod := range pods.Items {
+				var foundPod pod
+				assert.Conditionf(t, func() bool {
```
Comment (BupycHuk) [inline service/k8sclient/k8sclient_test.go:137]: let's fix this check

**Review by BupycHuk — APPROVED** [review]

File: service/k8sclient/k8sclient.go (hunk 12 above)
Comment (jprukner) [reply service/k8sclient/k8sclient.go:1222]: Ok, but I am confused. Last time I used, `new` I was told not to use it. I guess it was in another repo where not using `new` was more consistent then using it.

Comment (jprukner) [reply service/k8sclient/k8sclient.go]: Done

Comment (BupycHuk) [reply service/k8sclient/k8sclient.go]: Yeah, I know and we discussed it in pmm-go-backend. We will keep using `new` in `dbaas-controller` for now.
To start using `&` we have to update all places where we use `new`

**Comment by BupycHuk** [comment]
please update FB and move task to In QA

File: service/k8sclient/k8sclient_test.go (hunk 13 above)
Comment (jprukner) [reply service/k8sclient/k8sclient_test.go]: Done
